src/ImageSharp.Web/ImageSharpTagHelper.cs

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+using System.Text.Encodings.Web;
+using Microsoft.AspNetCore.Mvc.Razor.TagHelpers;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Mvc.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.Options;
+using SixLabors.ImageSharp.Processing;
+using SixLabors.ImageSharp.Web.Commands;
+using SixLabors.ImageSharp.Web.Middleware;
+using SixLabors.ImageSharp.Web.Processors;
+
+namespace SixLabors.ImageSharp.Web
+{
+    /// <summary>
+    /// A TagHelper implementation targeting &lt;img&gt; element that allows the automatic generation of HMAC protected processing commands.
+    /// </summary>
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + WidthAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + HeightAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + AnchorAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + RModeAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + XyAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + RColorAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + CompandAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + OrientAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + AutoOrientAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + FormatAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + BgColorAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    [HtmlTargetElement("img", Attributes = SrcAttributeName + "," + QualityAttributeName, TagStructure = TagStructure.WithoutEndTag)]
+    public class ImageSharpTagHelper : UrlResolutionTagHelper
+    {
+        private const string SrcAttributeName = "src";
+        private const string WidthAttributeName = ResizeWebProcessor.Width;
+        private const string HeightAttributeName = ResizeWebProcessor.Height;
```

Comment:
That was actually a conscious decision. My thought process was that it’s bad practice to use width/height attributes with unscaled content. Happy to be proven that this is a bad idea though.

Reply:
My main issue is that this will result in changing all existing image URLs to include the width/height (and HMAC token) commands by simply adding `@addTagHelper *, SixLabors.ImageSharp.Web` to your `_ViewImports.cshtml`. It will also process and cache these images, even though they will most likely already be the correct size (otherwise they would get scaled or distorted, which might even be done on purpose, e.g. when providing a high DPI version, upscaling, etc.).

Besides that, you might want to use a custom command to do the resizing (like Umbraco does with crop coordinates in the `CropWebProcessor`) and omit the width command from the URL while still explicitly setting the width attribute on the `img` tag. Another case might be when you disable the auto-orient feature and therefore have to swap the width/height commands.

You can avoid this by also using prefixes for these commands, so `<img src="test.jpg" imagesharp-width="500" imagesharp-height="500" />` will be rendered as `<img src="test.jpg?width=500&height=500" width="500" height="500" />`. This will still set the attributes, but only resize the image when explicitly set. This will also allow adding high DPI versions by using `<img src="test.jpg" width="250" height="250" imagesharp-width="500" imagesharp-height="500" />`.

Reply:
OK. You've convinced me. Explicit it is.

Reply:
Just to clarify the new behavior.

- We only resize if `imagesharp-width` or `imagesharp-height` are explicitly set. 
- We only set the width or height attributes if there are no attributes present.

Reply:
Correct, so:
- `<img src="sixlabors.imagesharp.web.svg" imagesharp-width="300" />` becomes:
  `<img src="sixlabors.imagesharp.web.png?width=300" width="300" />`
- `<img src="sixlabors.imagesharp.web.png" width="150" imagesharp-width="300" />` becomes:
  `<img src="sixlabors.imagesharp.web.png?width=300" width="150" />`

Reply:
The only thing that's not possible is only adding the width or height command without setting the attributes, although I can't think of any use cases where that would make sense...

I did notice the HMAC token doesn't include the query string/commands that are already in the src URL. It's probably a good idea to include that and provide a way to opt-in to adding the HMAC token using e.g `imagesharp-token="true"` or `imagesharp-hmac="true"`, in case no other tag helper attributes are use. This will allow easy HMAC token generation to URLs that already include ImageSharp commands.

Reply:
We don't need to do the first point since the cache key and HMAC are created only from the commands and not the raw querystring so additional unrecognized commands are ignored.

Regarding the second point. Yeah, we should protect all request coming to providers.

Adding `imagesharp-hmac` is enough now, values passed to the attribute are ignored.

Reply:
Looks like the query string/commands from the original `src` attribute aren't included in the HMAC token generation, as the following returns a 400 status code: `<img src="sixlabors.imagesharp.web.png?width=300" width="300" imagesharp-hmac />`. This currently makes the addition of the `imagesharp-hmac` attribute rather useless, as using any of the other `imagesharp-` attributes will already ensure the token is added when required.

If the commands are parsed from the `src` attribute, the tag helper could automatically append the HMAC token when a HMAC secret is configured and the URL contains known commands, without having to add any other `imagesharp-` attributes. This would make enabling HMAC processing protection on your existing site really easy, as you'd only have to ensure the tag helper is added to ensure all image URLs in your Razor views will have the correct HMAC tokens added.

This functionality could also be implemented for `picture/source` elements and `srcset` attributes. It might then be a good idea to separate this functionality into a different tag helper, so you can opt-in by adding e.g. `@addTagHelper SixLabors.ImageSharp.Web.TagHelpers.HmacTokenTagHelper, SixLabors.ImageSharp.Web` (and set the [`Order` property](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.razor.taghelpers.taghelper.order) to ensure it's processed after the `ImageTagHelper`). This avoids the URL processing hit in the `ImageTagHelper` and allows you to remove the dependency on `RequestAuthorizationUtilities` from that tag helper as well...

Another benefit would be that adding commands to the URL in the `src` attribute enables you to add ImageSharp commands in front of the ones added by the `imagesharp-` attributes, allowing you to change the order (which is important, as this impacts how the image is processed).

Reply:
Eh? Why the hell would someone write a command like that mixed in with the helper syntax? They do that and they’re on their own cos it’s bloody stupid

Reply:
Use the helper, don’t use the helper fine by me but don’t expect me to support stupid behaviour.

Reply:
I may not have explained it clearly enough 😇 There's a couple of benefits when separating the HMAC token generation into a separate tag helper:
1. You don't have to create your own tag helper (inherited from `ImageTagHelper`) to add a custom command (e.g. `imagesharp-cropcoordinates`) and have it included in the HMAC token, as you can then manually include that command in the image URL/`src` attribute;
2. The HMAC tokens can easily be added to other elements and attributes (e.g. `img[srcset]` and `source[srcset]`), without having to implement new tag helpers to generate the image URLs for these elements/attributes;
3. Existing image URLs that contain valid commands (e.g. returned from a CMS) can be augmented with additional commands and/or the HMAC token.

Maybe some code might help to explain it, so please have a look at these changes: https://github.com/SixLabors/ImageSharp.Web/commit/c72ee4a890759b5de4dd0357acaa51b059862dc8 (I've added some other suggestions to my fork, [see branch comparison](https://github.com/SixLabors/ImageSharp.Web/compare/js/taghelper...ronaldbarendse:ImageSharp.Web:js/taghelper)).

Reply:
@ronaldbarendse if you're still keen, I'd like to revisit your suggestion to split out the helper?

Reply:
OK. I ran with it. 

I have implemented pretty much everything you suggested with a few implementation tweaks. 

HMAC is now the separate helper you designed and requests with no commands no longer require HMAC.